### PR TITLE
feat(`grpo_trainer.py`): Variational Sequence-Level Soft Policy Optimization (VESPO)

### LINKED ISSUE #5196: feat: Variational Sequence-Level Soft Policy Optimization (VESPO)
### Feature request

Hi,

I'd like to propose adding VESPO (https://huggingface.co/papers/2602.10693) as a supported `loss_type` in `GRPOTrainer`

It's quite distinct from GRPO and closer to SAPO, with a smooth trust region.

- Instead of differentiating through the IS ratio, the gradient relies directly on the current logprobs (like vanilla policy gradient). The IS ratio, instead, is passed through a nonlinear function, detached, as a pure scaling factor.

- Advantages calculation is the same as GRPO. 
- Compatible with TIS/MIS.

The main highlight, imo, is nice/stable performance under bad policy staleness and training-inference mismatch, especially with MoEs.

&nbsp;

I wanted to know, first, if there is an interest to have this variant in TRL before opening a PR

### Motivation

I thought you might be interested in this one @qgallouedec  because they're getting quite nice results with sparse architectures specifically, even without routing replay, see some handpicked pics below:

Image

### Your contribution

yes

### PR DESCRIPTION
**Pull request by casinca**
# What does this PR do?

This PR implements the VESPO loss, resolve #5196

Official implementation: https://github.com/FloyedShen/VESPO/blob/main/recipe/vespo/code/core_algos.py
Paper: https://huggingface.co/papers/2602.10693

**Note:**
- The paper and the official implementation can have different variable names, to make things clearer:
    - c1 = k = α
    - c2 = lambda
    
- Docstrings/comments are a mix of official impl and my writing.

&nbsp;

**Alternative options:**

- Currently VESPO has 4 hparams `k_pos`, `lambda_pos`, `k_neg`, `lambda_neg` but I could reduce with 2 tuples of 2 floats eg: lambdas (pos, neg) if it's better.
- Original impl also returns for metrics `w_seq`. I can include it in metrics, but this would force me to return a tuple in `get_gamma_weights` or remove `@staticmethod`. Not sure here what's the preference.

&nbsp;

For efficiency, the TRL VESPO implementation is slightly different than the official one. It's ~25% faster per call on gpu, and tested for equivalence.

```bash
With importance_sampling_ratio:
-----------------------------------------------------------------
B x T         TRL_VESPO (ms)    OG_VESPO (ms)     Faster
-----------------------------------------------------------------
8 x 128         0.4290          0.5301          TRL_VESPO (1.24x)
16 x 256        0.4281          0.5302          TRL_VESPO (1.24x)
32 x 512        0.4283          0.5299          TRL_VESPO (1.24x)
64 x 512        0.4284          0.5294          TRL_VESPO (1.24x)
128 x 512       0.4286          0.5322          TRL_VESPO (1.24x)
32 x 1024       0.4473          0.5313          TRL_VESPO (1.19x)
64 x 1024       0.4285          0.5360          TRL_VESPO (1.25x)
128 x 1024      0.4240          0.5203          TRL_VESPO (1.23x)
-----------------------------------------------------------------
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Adds a new loss formulation to `GRPOTrainer` that changes how gradients are scaled and how vLLM importance-sampling correction is applied, so training behavior can shift for users selecting `loss_type="vespo"`. Scope is contained to a new loss branch plus config/docs/tests, but it touches core loss computation.
> 
> **Overview**
> Adds **VESPO** (`loss_type="vespo"`) to `GRPOTrainer`, including a new `get_gamma_weights` helper that computes detached, advantage-sign-dependent Gamma weights from sequence-level importance ratios (optionally incorporating vLLM TIS/MIS correction in log space) and uses them in the loss.
> 
> Extends `GRPOConfig` with four new hyperparameters (`vespo_k_pos`, `vespo_lambda_pos`, `vespo_k_neg`, `vespo_lambda_neg`) and updates validation/behavior so VESPO warns about `importance_sampling_level` usage, restricts vLLM correction modes to `token_truncate`/`token_mask`, and skips the generic per-token vLLM correction multiplier for VESPO.
> 
> Updates docs (`paper_index.md`) to add a VESPO paper entry/config snippet and expands the GRPO loss-type test matrix to include `vespo`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit c5d0d50036fc2b6a2f1064b6ccabc790ce15a15a. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -2076,6 +2047,57 @@ def get_off_policy_mask(
...
+                with w = sequence-level importance sampling ratio
+        note: we will compute φ(w) in log space
+
+        φ(w) is detached via @torch.no_grad(), only acts as gradient scaling coefficient
+
+        VESPO loss =  -φ(w) × A × log_prob, gradient naturally gives φ(w) × A × ∇log π
+        """
+        # reducing clamp range directly to log(1e-8) ~ -18.42, to avoid recomputing log_w=log(w.clamp(min=1e-8)) later
+        # This is solely for matching truthfully the original implementation, otherwise keeping -20 could be fine.
+        lower_clamp = math.log(1e-8)
```
Comment (casinca) [inline trl/trainer/grpo_trainer.py]: This is an optimization over the og impl, but this requires importing `math`, just for the log.
Otherwise I'd need to hardcode or convert from a torch scalar with torch.log().

This avoids recomputing a log op in torch.log(w_seq) when we already have log_w.

I opened an issue in OP because not sure if -18.42 is really making a difference over -20: https://github.com/FloyedShen/VESPO/issues/6

Anyway this is to stay true to the official impl

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -2076,6 +2047,57 @@ def get_off_policy_mask(
...
+
+        # Sequence-level log ratio Σ log(π_θ/π_old) (not a mean like for `log_importance_weights`)
+        log_ratio_clamped = torch.clamp(log_ratio_per_token, -20.0, 20.0)
+        seq_log_ratio = torch.sum(log_ratio_clamped * mask, dim=-1, keepdim=True)  # (B, 1)
+
+        # Apply token-level TIS or MIS correction (in log space)
+        if importance_sampling_ratio is not None:
```
Comment (casinca) [inline trl/trainer/grpo_trainer.py]: token-level MIS is possible, so I didn't forbid it. but OG impl is token-level TIS tested.

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -2186,7 +2220,7 @@ def _compute_loss(self, model, inputs):
         if entropy_mask is not None:
             per_token_loss = per_token_loss * entropy_mask
 
-        if self.use_vllm and self.vllm_importance_sampling_correction:
+        if self.use_vllm and self.vllm_importance_sampling_correction and self.loss_type != "vespo":
```
Comment (casinca) [inline trl/trainer/grpo_trainer.py]: `importance_sampling_ratio` has to be applied separately for VESPO, in `get_gamma_weights` , hence preventing it being applied here.

File: trl/trainer/grpo_trainer.py [hunk 4]
```diff
@@ -2205,7 +2239,7 @@ def _compute_loss(self, model, inputs):
             loss = (per_token_loss * mask).sum() / (per_token_loss.size(0) * self.max_completion_length)
             normalizer = self.current_gradient_accumulation_steps if mode == "train" else 1.0  # no accum in eval
             loss = loss / normalizer
-        elif self.loss_type in ["cispo", "dapo"]:
+        elif self.loss_type in ["cispo", "dapo", "vespo"]:
             normalizer = inputs["num_items_in_batch"] / self.accelerator.num_processes
             loss = (per_token_loss * mask).sum() / normalizer
```
Comment (casinca) [inline trl/trainer/grpo_trainer.py]: Default aggregation per the og impl, is token-mean

**Comment by casinca** [comment]
I owe some better explanations to facilitate the review concerning importing `math` for `lower_clamp = math.log(1e-8)` in `get_gamma_weights`

From the original implementation below, the author is recomputing the `log_w_seq` from `w_seq` but we already have the log from `seq_log_ratio_clampled`. The only difference for recomputing, is the range of the min clamp being reduced to `min=1e-8`.

image

&nbsp;

In order to avoid a 2nd log op in TRL, I'm directly clamping in logspace `log_w_seq = torch.clamp(seq_log_ratio, lower_clamp, 20.0)` once. Which ends up being the same.

This is solely to follow the original implementation, otherwise I'm not really sure if reducing from  $e^{-20}$ to  $e^{-18.42}$ (ie $e^{log(1e-8)}$) is important. I had opened an issue in OP for this: https://github.com/FloyedShen/VESPO/issues/6

If keeping the original logic and importing `math` is problematic, alternative could be to hardcode log(1e-8) or retrieved from a tensor `torch.log(torch.tensor(1e-8))`

**Comment by casinca** [comment]
hey @FloyedShen , feel free to share any thoughts on this TRL implementation.

**Comment by FloyedShen** [comment]
Hi @casinca, sorry for the late reply — been swamped with work lately 😅

Thanks for this great contribution! I'm one of the authors of VESPO. I took some time to go through your implementation and ran a quick validation experiment comparing VESPO against the GRPO baseline.

**Setup:** Qwen3-4B-Base, DAPO-Math-17k, 8×H20 GPUs, 8 generations, 8 steps per generation, lr=1e-6, max_completion_length=8192.

The results look solid and are consistent with what we see in our verl-based experiments — VESPO shows improved reward over GRPO as training progresses, with notably more stable gradient norms and better entropy retention throughout training. Everything checks out on my end!

Screenshot 2026-03-14 at 01 52 02

Full training logs: https://wandb.ai/brain-cog/vespo_trl

Thanks again for the clean implementation and the thorough benchmarking 👍

**Comment by casinca** [comment]
Np, thanks for the feedback and for taking the time to test, appreciate it.

File: trl/trainer/grpo_trainer.py [hunk 5]
```diff
@@ -578,6 +579,19 @@ def __init__(
                 "paper's setup."
             )
 
+        if args.loss_type == "vespo" and args.importance_sampling_level != "token":
+            logger.warning(
+                "VESPO computes sequence-level importance weights internally. `importance_sampling_level` should be "
+                "set to `'token'` (the default)."
+            )
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:586]: I just added a warning here
